### PR TITLE
Use dagMap to improve plugin ordering

### PR DESCRIPTION
**Pull request by sarupbanskota**
Works towards #5623, Improving #5735 

(Thanks to Nathan for providing me much needed guidance!)

| Q                        | A 
| ------------------------ | ---
| Patch: Bug Fix?          | 
| Major: Breaking Change?  | 
| Minor: New Feature?      | yes
| Deprecations?            | 
| Spec Compliancy?         | 
| Tests Added/Pass?        | 
| Fixed Tickets            | Towards fixing #5623, Improves #5735 
| License                  | MIT
| Doc PR                   | 
| Dependency Changes       | yes

### CODE REVIEW

**Comment by xtuc** [comment]
At first glance it looks awesome, thanks for this PR :+1:

Going to look in more detail later on.

File: package.json [hunk 1]
```diff
@@ -57,5 +57,8 @@
     ],
     "sourceMap": false,
     "instrument": false
+  },
+  "dependencies": {
+    "dag-map": "^2.0.2"
```
Comment (hzoo) [inline package.json]: This would need to go into babel-core since it's being used there. The top level package.json doesn't apply to any specific package and used for devDeps

**Comment by hzoo** [comment]
You can also just target my PR instead of targeting 7.0

**Comment by sarupbanskota** [comment]
Sure - I'll be back at this later tonight! Need to look at why the tests broke too - strange

**Comment by hzoo** [comment]
I forgot to push some code from earlier - https://github.com/babel/babel/compare/plugin-ordering...plugin-ordering-2?expand=1. It might fix some of the tests

**Comment by sarupbanskota** [comment]
@hzoo I think you can pull this in, sounds like ncantelmi will have a PR to fix broken tests on this PR, and better dependency handling

**Comment by hzoo** [comment]
Can you look at my branch again? I removed the error check https://github.com/babel/babel/compare/plugin-ordering...plugin-ordering-2?expand=1 and the dag-map dependency is in the incorrect place

**Comment by nathanhammond** [comment]
> I removed the error check -@hzoo

Removing that error check hides cyclic dependencies:
```
1 = { capabilities: [A, B] }
2 = { capabilities: [A], dependencies: [B, C] }
3 = { capabilities: [B], dependencies: [A] }
4 = { capabilities: [C], dependencies: [A] }

A: 2
B: 3
C: 4

graph.add(1)
graph.add(2, [3,4])
graph.add(3, [2])
graph.add(4, [2])
```

This is otherwise resolveable, but we have lost the metadata to properly sort `1`. Getting a valid result here is not possible to guarantee for all scenarios even with "stability" guarantees in the topological sort because math (trivial examples like this look solvable with stability, but I assure you that real-world graphs are impossible).

In other words, you can't have multiple plugins provide the same capability. Given that, retaining the error is correct.

File: packages/babel-core/src/config/option-manager.js [hunk 2]
```diff
@@ -185,7 +185,18 @@ class OptionManager {
...
+      const uniquePlugins = plugins
+        .reduce((accumulator, pluginTuple) => {
+          const plugin = pluginTuple[0];
+          if (!accumulator[plugin.key]) {
+            accumulator[plugin.key] = pluginTuple;
+          }
+          return accumulator;
+        }, {});
+
+      plugins = sortPlugins(Object.values(uniquePlugins));
```
Comment (nathanhammond) [inline packages/babel-core/src/config/option-manager.js]: Failing in less than Node 6 because of `Object.values`.
